compaction/compaction.cc

```diff
@@ -137,14 +139,28 @@ std::string_view to_string(compaction_type_options::scrub::quarantine_mode quara
 
 static api::timestamp_type get_max_purgeable_timestamp(const table_state& table_s, sstable_set::incremental_selector& selector,
         const std::unordered_set<shared_sstable>& compacting_set, const dht::decorated_key& dk, uint64_t& bloom_filter_checks,
-        const api::timestamp_type compacting_max_timestamp) {
+        const api::timestamp_type compacting_max_timestamp, is_shadowable is_shadowable) {
     if (!table_s.tombstone_gc_enabled()) [[unlikely]] {
         return api::min_timestamp;
     }
 
     auto timestamp = api::max_timestamp;
-    auto memtable_min_timestamp = table_s.min_memtable_timestamp();
-    // Use memtable timestamp if it contains data older than the sstables being compacted,
+    api::timestamp_type memtable_min_timestamp;
+    if (is_shadowable) {
+        // For shadowable tombstones, check the minimum live row_marker timestamp
+        // as it shadows any older shadowable tombstones, exposing live cells with
+        // timestamps potentially lower than the shadowable tombstone (those
+        // are tracked in the min_memtable_live_timestamp)
+        // See https://github.com/scylladb/scylladb/issues/20424
+        memtable_min_timestamp = table_s.min_memtable_live_row_marker_timestamp();
+    } else {
+        // For regular tombstones, check the minimum live data timestamp.
+        // We don't care if the purged tombstones shadows dead data
```

Comment:
Perhaps instead: When making a decision whether to purge an sstable tombstone or not, we don't need to coordinate with memtable tombstone timestamps since even a backdated memtable tombstone can not resurrect any dead data in the sstable.

Reply:
> Perhaps instead: When making a decision whether to purge an sstable tombstone or not, we don't need to coordinate with memtable tombstone timestamps since even a backdated memtable tombstone can not resurrect any dead data in the sstable.

There are no such things as sstable tombstones and memtable tombstones.
And the logic applies also for dead data in other sstables, not just the memtable (though the check for other sstables is implemented a bit later in this function).

How about: Even if purgeable tombstones shadow dead data in the memtable, it's ok to purge them, since "resurrecting" the already-dead data will no have effect, as they are already dead.

Reply:
Perhaps "purgeable tombstones" is a well understood term, but to me this sounds less clear.

Purgeable from where? sstables, presumably, so I called these tombstones, stored in sstables, "sstable tombstones", and this is where you lost me. 

I give up, you new comment is not much more clear to me, but my bikeshedding alarm is off, so I resolve.